AnimatedPlaceholder: Replace sliding with typing animation

alshakero opened the pull request:
This PR replaces the sliding animation with typing animation in `AnimatedPlaceholder` component. I also propose a beautiful (imo) fade away animation (from left to right) that I accidentally stumbled upon.

Video before: https://d.pr/v/1shTAb
Video after: https://d.pr/v/1fc8xL

## Testing

Navigate to `/gutenboarding` and you should see the difference right away.

Fixes #40387

sirreal commented:
I'd been thinking about this and was really expecting something to actually manipulate text.

https://www.npmjs.com/package/typed.js can provide some inspiration, but a hook React might be a nice interface for this in our project:

```tsx
const typedValue = useTyper( [ 'string1', ...strings ] );
<div>{ typedValue }</div>;
```

I think that would provide a few benefits:

- Use it anywhere you want text, e.g. a `placeholder`.
- Not tie this to a particular design/CSS
- We could provide some jitter so typing isn't irregular.

What do you think?

alshakero commented:
Funny you say that! I created the hook then trashed it. I agree with the idea for sure, I trashed it because I wanted to minimize the diff size. I'll recreate it.

alshakero commented:
I implemented the hook and reverted the previous changes.

For easier reviewing, I suggest reviewing using full-pr-diff mode (as opposed to the latest commit diff) because I reverted some changes and didn't want to force push.

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,49 @@
+	delayBetweenCharacters = 75,
+	delayBetweenWords = 750
+) {
+	const [ indices, setIndices ] = useState( { char: 0, word: 0 } );
```
Being able to break state down into units rather than shoving everything into an object is one of the big advantages of `useState` over class component's `setState`. Shall we keep them separate here?

alshakero replied:
I agree but I joined them for a reason. Whenever you update the state a setTimeout will be called because your update will change the deps values.

So in [this line](https://github.com/Automattic/wp-calypso/pull/40431/files#diff-4b30f36c7b54c2125d9a0f7663b3a335R38), I update both indices (char = 0, word++). If the states were separate, the first update will probably trigger an update => set a timeout and mess things up. Having one state means a linear predictable stream of events.

razvanpapadopol commented:
Awesome looking animation! 🏆 

Is there a way to use some reverse (but faster) effect when fading out the previous word like it's in the prototype? https://cloudup.com/cN9QY5EzYyq
I like the current dynamic way of transitioning (instead of slowly fading out using opacity) but seems just a little too abrupt. cc @dubielzyk

sirreal commented:
> Is there a way to use some reverse (but faster) effect when fading out the previous word like it's in the prototype? https://cloudup.com/cN9QY5EzYyq

I like reversing the word as well 👍 

I don't mind if we implement that here or add it as an enhancement after the hook is merged. I think we'll be tweaking the delays as will, it seems very fast to me.

alshakero commented:
I adopted [Dan Abramov's method of creating an interval hook](https://overreacted.io/making-setinterval-declarative-with-react-hooks/) then cleaned up my hook and implemented the deleting animation.

sirreal commented:
> I adopted [Dan Abramov's method of creating an interval hook](https://overreacted.io/making-setinterval-declarative-with-react-hooks/) then cleaned up my hook and implemented the deleting animation.

We have `use-interval` already sitting around somewhere for reuse 🙂

alshakero commented:
> We have use-interval already sitting around somewhere for reuse 🙂

Lovely!

alshakero commented:
> One concern is that this doesn't behave well if it receives an empty array of texts. Can we protect against that?

Addressed

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,61 @@
+) {
+	const [ charIndex, setCharIndex ] = useState( 0 );
+	const [ wordIndex, setWordIndex ] = useState( 0 );
+	const [ mode, setMode ] = useState( 'TYPING' );
```
Shall we restrict this?

```
// Add above hook definition
type TyperMode = 'TYPING' | 'DELETING';
```

```suggestion
	const [ mode, setMode ] = useState< TyperMode >( 'TYPING' );
```

alshakero replied:
Neat. Still not used to TS pampering.

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,61 @@
+	texts: Array< string >,
+	enabled = true,
+	delayBetweenCharacters = 120,
+	delayBetweenWords = 1200
```
What do you think about all the parameters listed vs. putting some, like these configuration values, into an object?

alshakero replied:
I felt this was a little over-complicated for its value but the hook API looks much better now. Adopted.

alshakero commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,61 @@
+	enabled = true,
+	delayBetweenCharacters = 120,
+	delayBetweenWords = 1200
+) {
```
Solid point.

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+type AllTyperOptions = {
+	delayBetweenCharacters: number;
+	delayBetweenWords: number;
+};
```
I'd expect an interface for this. Simple `Options` name is fine for this local type. We'll leave the properties non-optional which works nicely with other mapped types.

```tsx
interface Options {
	delayBetweenCharacters: number;
	delayBetweenWords: number;
};
```

I've also really liked documenting the properties right in the interface:

```suggestion
interface Options {
	/**
	 * The typing delay time in ms between two characters.
	 *
	 * @default 120
	 */
	delayBetweenCharacters: number;
	
	/**
	 * The typing delay in ms between two words.
	 *
	 * @default 1200
	 */
	delayBetweenWords: number;
};
```

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+	| Pick< AllTyperOptions, 'delayBetweenCharacters' >
+	| Pick< AllTyperOptions, 'delayBetweenWords' >
+	| AllTyperOptions;
+
```
We don't get much value from this and it's pretty complicated. If someone really wants to pass `{}` for options that's not a problem. The types will say options isn't required anyways.

```suggestion
```

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+	| Pick< AllTyperOptions, 'delayBetweenWords' >
+	| AllTyperOptions;
+
+const defaultOptions: AllTyperOptions = { delayBetweenCharacters: 120, delayBetweenWords: 1200 };
```
Use `Options`:

```tsx
const defaultOptions: Options = { delayBetweenCharacters: 120, delayBetweenWords: 1200 };
```

We can also prevent it from being modified:

```tsx
const defaultOptions: Readonly< Options > = { delayBetweenCharacters: 120, delayBetweenWords: 1200 };
```

Finally, I'd probably use CAPITAL_SNAKE_CASE for this constant:

```suggestion
const DEFAULT_OPTIONS: Readonly< Options > = { delayBetweenCharacters: 120, delayBetweenWords: 1200 };
```

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+ * @param enabled Whether the animation is enabled
+ * @param options Animation options
+ * @param options.delayBetweenCharacters The typing delay time between two charactors. Default is 120ms
+ * @param options.delayBetweenWords The typing delay time between two words. Default is 1200ms
```
Optionally, we can leave these documented in the interface.

```suggestion
```

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+export default function useTyper(
+	words: Array< string >,
+	enabled = true,
+	options: TyperOptions = defaultOptions
```
We'll be merging anyways, we don't need to set to the default object then merge it with itself:

```tsx
	options?: Options
```

Here, we'll use the `Partial` mapped type to make all properties optional.

```suggestion
	options?: Partial< Options >
```

alshakero replied:
TIL: you can spread `undefined` 😮 

```js
{...undefined} == {}
```

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+	const [ wordIndex, setWordIndex ] = useState( 0 );
+	const [ mode, setMode ] = useState< TyperMode >( 'TYPING' );
+
+	const populatedOptions: AllTyperOptions = { ...defaultOptions, ...options };
```
```suggestion
	const populatedOptions: Options = { ...DEFAULT_OPTIONS, ...options };
```

sirreal commented on client/landing/gutenboarding/hooks/use-typer.ts:
```diff
@@ -0,0 +1,84 @@
+ */
+export default function useTyper(
+	words: Array< string >,
+	enabled = true,
```
This is probably inferred, but I like to keep these explicit:

```suggestion
	enabled: boolean = true,
```

alshakero replied:
TS doesn't like this. 

![image](https://user-images.githubusercontent.com/17054134/77745275-473f9a80-701b-11ea-8579-20d40bdd2107.png)
